### DISCUSSION TITLE
Weird behavior on a toy dataset

### DISCUSSION THREAD

**Discussion by mys007** [voxelmorph/voxelmorph#286] [Q&A]
I'm trying to understand the behavior of VoxelMorph (PyTorch version, current master) by overfitting on a 2D toy dataset. The dataset consists of a single image pair with 2 rectangles which overlap to a certain degree (AFAIK VoxelMorph cannot deal with non-overlapping ones because there is no multi-scale loss). The generator is simple:
```
def toy_square_reg(batch_size=1, **kwargs):
    empty = np.zeros((batch_size, 512, 512, 1))
    img1 = np.zeros((batch_size, 512, 512, 1))
    img2 = np.zeros((batch_size, 512, 512, 1))
    img1[:,100:150, 200:300] = 1
    img1[:,300:350, 200:300] = 2
    img2[:,100:150, 290:320] = 1
    img2[:,300:350, 200:260] = 2
    while True:
        invols  = [img1, img2]
        outvols = [img2, img1]
        yield (invols, outvols)     
```

I use default parameters but no regularization in a wish to arrive at 0 loss - but it does not converge. See this [video](https://github.com/voxelmorph/voxelmorph/files/6197393/voxelmorph_toy_lambda0.zip) for step-by-step behavior of the fixed image, the moving image and the deformation field. 

There are two issues:

1) One of the boxes is subject to an unexpected deformation. There seems to be some kind of "link" between the two boxes, which intuitively should not be if there is no regularization.
2) In general, boxes tend to get "streaks" or "mustaches" at corners until they have converged. I have observed these artifacts also on real-world data where VoxelMorph tends to produce van-Gogh-like artifacts.

Note that the following observations hold also when L2 gradient regularizer is used. Alas with increasingly strong `lambda` the network sort of gets it right but one see it really fights hard. For example, see this [video](https://github.com/voxelmorph/voxelmorph/files/6197391/voxelmorph_toy_lambda0.1.mkv.zip) for `lambda=0.1`.

I would be very happy for any insights, thanks!

**Comment by adalca** [comment]
Hi @mys007,

interesting test case -- can you tell me which model you used? I wonder if it's an issue of optimization (with lambda=0, it's easier to just collapse the squares, since this gives you matching 0s, rather than try to match them) or maybe an issue of the model not having enough FoV. I think lambda=0.1 is a more interesting test -- can you show me the loss convergence curve for that one?

**Comment by mys007** [comment]
Thanks for your quick reply!

> can you tell me which model you used?

I use the default one, i.e. fresh `vxm.networks.VxmDense` with the default hyperparameters.

> with lambda=0, it's easier to just collapse the squares, since this gives you matching 0s, rather than try to match them

I agree that if for some reason the fixed image rectangles cannot be aligned with, "deleting" the moving rectangles is optimal. But if they can, wouldn't a black hole collapse of the moving rectangles still incurs non-zero penalty at the places where the rectangles in the fixed image lie (since MSE is used), thus being suboptimal?

> can you show me the loss convergence curve (...) lambda=0.1

Here it is but for a different progression (turns out the convergence is not reproducible likely due to non-fixed seeds etc), see the [new video](https://github.com/voxelmorph/voxelmorph/files/6200733/voxelmorph_toy_lambda0.1_run2.mkv.zip).

<img src="https://user-images.githubusercontent.com/5921083/112388921-2b390c80-8cf4-11eb-9d57-34b59724eb13.png" height="300">

**Comment by adalca** [comment]
> I'm not sure, wouldn't a black hole collapse of the moving rectangles still incurs non-zero penalty at the places where the rectangles in the fixed image lie (since MSE is used)?

sorry, i meant an easy *local* solution. You're right, globally your solution is better. 
But there is a local minima by collapsing the squares.

It may be (I'm hypothesizing) that by turning of the lambda, it finds this local minima faster and easier. 
With *some* lambda value it has to work through smoother solutions and perhaps this helps since collapsing a square would require non-uniform transforms.

Anyway, I tried your example and it worked ok for me, more or less -- here's a [colab](https://colab.research.google.com/drive/1U62UqeRalIQPUP5bIX8Unmydms95B79A?usp=sharing). I guess it's like you say, it's trying hard, but that's not all that surprising, it's just an optimization problem at this point? 

This balance/tradeoff of minimas will likely change if you move the squares closer or change the SGD LR so that it can take bigger crazier steps.

I'd recommend you play with 64x64 images instead of 512x512 :)

**Comment by mys007** [comment]
Thanks again for the swift response and trying it out! It's TF on colab but I assume the implementation is identical to PyTorch.

I agree that it's an optimization problem but many components have an effect. One of them is the data loss: with MSE, it looks like it's a matter of the ratio of overlap and non-overlap (or alternatively: it also works when I downweigh the loss term on the black background) whether the boxes get aligned or one of them "disappears".

But I've also found out that the integration hides insights. If I turn it off (`--int-steps 0`), one can see for some reason the network massively struggles to vacate the black areas - even without regularization, which is really weird [video](https://github.com/voxelmorph/voxelmorph/files/6204388/voxelmorph_toy_lambda0_iter0.zip). With added regularization this stabilizes a bit but tends to stuck in non-optimal equilibriums. 

So I'm wondering what effect the choice of the network has. For example, there seems to also be correlation between fields in unrelated part of the image (ie between the two boxes), demonstrated as synchronized behavior jumps during optimization. This is perhaps due overly high field of view of the UNet.

Well, I guess this turned out to be to difficult to disentangle quickly and it also doesn't seem like a classical bug in the code, so perhaps it's a subject to some future paper;). I guess it would be interesting to see how classic non-DL algorithms can deal with this image pair and how hyper-parameters affect this, but I don't have time for this now:).

Anyway, thanks for your help!

**Comment by adalca** [comment]
> Thanks again for the swift response and trying it out! It's TF on colab but I assume the implementation is identical to PyTorch.

Indeed, they should be very similar (I worry about claiming "identical" :) )

> I agree that it's an optimization problem but many components have an effect. One of them is the data loss: with MSE, it looks like it's a matter of the ratio of overlap and non-overlap (or alternatively: it also works when I downweigh the loss term on the black background) whether the boxes get aligned or one of them "disappears".

Yeah, this makes sense. 

> But I've also found out that the integration hides insights. If I turn it off (`--int-steps 0`), one can see for some reason the network massively struggles to vacate the black areas - even without regularization, which is really weird [video](https://github.com/voxelmorph/voxelmorph/files/6204388/voxelmorph_toy_lambda0_iter0.zip). With added regularization this stabilizes a bit but tends to stuck in non-optimal equilibriums.

My intuition is that no regularization means harder to predict in a way, as it seems like you'd be more susceptive to initializations. I suspect the regularization doesn't just help the loss be better behaved, but in some sense the optimization of the field as well.
 
> So I'm wondering what effect the choice of the network has. For example, there seems to also be correlation between fields in unrelated part of the image (ie between the two boxes), demonstrated as synchronized behavior jumps during optimization. This is perhaps due overly high field of view of the UNet.

Well you could take the network out of the equation and use the InstanceDense (this is a "network" whose only learnable parameters represent the field itself). That would essentially eliminate the network and do classical pairwise optimization.

> Well, I guess this turned out to be to difficult to disentangle quickly and it also doesn't seem like a classical bug in the code, so perhaps it's a subject to some future paper;). I guess it would be interesting to see how classic non-DL algorithms can deal with this image pair and how hyper-parameters affect this, but I don't have time for this now:).

Yup it sounds like standard optimization questions. BTW, one thing to try, since `vxm` is all about learning to approximate the registration output, is to feed voxelmorph image pairs similar to the one you've built, but slightly different each time. E.g. at each iteration, initialize one image with two boxes randomly placed, and another image with those boxes shifted. Seeing multiple examples will slowly teach `vxm` about this sort of exercise, and it may well perform better than if it trains on a single instance (what you've been doing).

> Anyway, thanks for your help!

Glad to help. I'm going to close this but feel free to comment if you try more things. I might also move this `issue` to a `discussion` since it turns out to not really be an issue it seems, but more just ... research :)

**Reply by mys007** [reply]
Hi, thanks for your insights. The pointer to InstanceDense was a pretty good one (I coded a PyTorch variant of this).

It turned out the registration works pretty well if the network is completely removed from the way. Even with the original MSE loss (no downweighting of the background): see this 
[video](https://github.com/voxelmorph/voxelmorph/files/6213618/field_int0_lam0.01_lr1.mkv.zip): 0 integration steps and lambda=0.01. The fact that one of the boxes does not fully converge is related purely to the L2 grad regularization which blocks the total compression, with lower regularization it works perfectly. 

There are two observations:

1) The "streaks" or "mustaches" at corners I've mentioned in the original post are apparently caused by the integration, see this 
[video](https://github.com/voxelmorph/voxelmorph/files/6213630/field_int7_lam0.01_lr1.mkv.zip) with 7 integration steps. So I believe it's either caused by the scaling and squaring method principally or they way it's implemented (perhaps due to rotational variance or so?).

2) The UNet incurs some effects which prevented the registration from working well. As different CNNs have different "priors" (as shown eg in the famous [Deep Prior](https://arxiv.org/abs/1711.10925) paper), perhaps plain old UNets are not the best option for registration problems in general...

I guess both of those issues point to some interesting directions.

**Reply by adalca** [reply]
Thanks for the update, all good stuff!

Intuitively I think it makes sense that the integration would cause some issue in this case, since it would drive registration towards more regular Jacobians, so it's more constrained and limits what's possible in the optimization. I wonder what happens if you initialize a network with the result from 0 integration steps, and start training with 7 integration steps. It might do ok, now that it's initialized in a more reasonble direction.

It also definitely makes sense that the network itself imposes priors -- something we've noted in the original vxm, but didn't really deeply analyze. I think this is actually a good thing in most situations (at least that was our interpretation) -- you don't often want to move white boxes on black backgrounds in practice :) . But I agree that there are some corner cases for which the unet is not expressive enough. Of course, the unet was just one simple architectural choice for the first vxm experiments, but there's nothing limiting vxm to use unets!
